From c92add3648c222089ae9844da5fa760ae59a4ca1 Mon Sep 17 00:00:00 2001
From: dasdgw <michifrank2@gmx.de>
Date: Thu, 20 Jun 2024 14:32:23 +0200
Subject: [PATCH 2/2] map sx88v2 remote control EV_MSC to EV_KEY

the sx88v2 is reporting events of type EV_MSC.
But rcinput filters for EV_KEY.
Remap all EV_MSC events to unused EV_KEY events.
---
 configure.ac           |   4 ++
 lib/driver/rcinput.cpp | 104 ++++++++++++++++++++++++++++++++++++++++-
 2 files changed, 107 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index b11f145..8993c51 100644
--- a/configure.ac
+++ b/configure.ac
@@ -617,6 +617,10 @@ if test "$BOXTYPE" == "sf8008" -o "$BOXTYPE" == "sf8008m" -o "$BOXTYPE" == "sx98
 	AC_DEFINE(FORCE_ALPHABLENDING_ACCELERATION, 1,[define when the framebuffer acceleration has alphablending support, but detection slow down all])
 fi
 
+if test "$BOXTYPE" == "sx88v2"; then
+	AC_DEFINE(SX88V2RC, 1,[enable octagon sx rcu 03 extra buttons])
+fi
+
 if test "$BOXTYPE" == "et4x00" -o "$BOXTYPE" == "et5x00" -o "$BOXTYPE" == "et6x00" -o "$BOXTYPE" == "et7x00" -o "$BOXTYPE" == "et9x00" -o "$BOXTYPE" == "et8000" -o "$BOXTYPE" == "et8500" -o "$BOXTYPE" == "et8500s" -o "$BOXTYPE" == "et10000"; then
 	AC_DEFINE(KEY_PLAY_ACTUALLY_IS_KEY_PLAYPAUSE, 1,[define when rc sends a KEY_PLAY event for its KEY_PLAYPAUSE key])
 	AC_DEFINE(KEY_F3_TO_KEY_LIST, 1,[define when rc sends a KEY_F3 event for its KEY_LIST key])
diff --git a/lib/driver/rcinput.cpp b/lib/driver/rcinput.cpp
index 31b096c..834d569 100644
--- a/lib/driver/rcinput.cpp
+++ b/lib/driver/rcinput.cpp
@@ -13,10 +13,100 @@
 #include <lib/base/init_num.h>
 #include <lib/driver/input_fake.h>
 
+#define SX88V2_TV_PWR   0x7f8f10e
+#define SX88V2_TV_VOLM  0x7f8f20d
+#define SX88V2_TV_VOLP  0x7f8f30c
+#define SX88V2_TV_TVAV  0x7f8f40b
+
+#define SX88V2_IPTV     0x7f849b6
+#define SX88V2_SAT      0x7f8659a
+
+#define SX88V2_PREVIOUS 0x7f8629d
+#define SX88V2_NEXT     0x7f8609f
+
+#define SX88V2_NETFLIX  0x7f853ac
+#define SX88V2_KODI     0x7f857a8
+#define SX88V2_YOUTUBE  0x7f85da2
+#define SX88V2_WEBTV    0x7f8ea15
+
+static unsigned short int pressedKey;
+
 void eRCDeviceInputDev::handleCode(long rccode)
 {
 	struct input_event *ev = (struct input_event *)rccode;
-	eDebug("[eRCDeviceInputDev] ==> BEFORE check for evtype: %x %x %x", ev->value, ev->code, ev->type);
+
+#if SX88V2RC
+	bool skipKey = false;
+	eDebug("[eRCDeviceInputDev] ==> BEFORE check for evtype: value: %x code: %x type: %x", ev->value, ev->code, ev->type);
+	if ((ev->code == 0x4) && (ev->type == EV_MSC)) {
+		switch (ev->value)
+		{
+		case SX88V2_KODI:
+			//eDebug("[eRCDeviceInputDev] kodi key detected");
+			pressedKey = KEY_PLAYER;
+			break;
+		case SX88V2_NETFLIX:
+			pressedKey = KEY_MP3;
+			break;
+		case SX88V2_YOUTUBE:
+			pressedKey = KEY_KEYBOARD;
+			break;
+		case SX88V2_WEBTV:
+			pressedKey = KEY_VOD;
+			break;
+		case SX88V2_TV_PWR:
+			pressedKey = KEY_ZOOM;
+			break;
+		case SX88V2_TV_VOLM:
+			pressedKey = KEY_F10;
+			break;
+		case SX88V2_TV_VOLP:
+			pressedKey = KEY_F6;
+			break;
+		case SX88V2_TV_TVAV:
+			pressedKey = KEY_MEDIA;
+			break;
+		case SX88V2_IPTV:
+			pressedKey = KEY_WWW;
+			break;
+		case SX88V2_SAT:
+			pressedKey = KEY_ARCHIVE;
+			break;
+		case SX88V2_PREVIOUS:
+			pressedKey = KEY_PREVIOUS;
+			break;
+		case SX88V2_NEXT:
+			pressedKey = KEY_NEXT;
+			break;
+		default:
+			skipKey = true;
+			break;
+
+		}
+		// press
+		//input->keyPressed(eRCKey(this, pressedKey, 0)); /*emit*/
+		if (skipKey == false)
+		{
+			ev->value = 0x1;
+			ev->code = pressedKey;
+			ev->type = EV_KEY;
+		}
+	}
+
+	if ((ev->value == 0x0) && (ev->code == 0x0) && (ev->type == 0x0)) {
+		if (pressedKey != 0x0) {
+			//eDebug("[eRCDeviceInputDev] releasing last key");
+			// release
+			//input->keyPressed(eRCKey(this, pressedKey, eRCKey::flagBreak)); /*emit*/
+			ev->value = 0x0;
+			ev->code = pressedKey;
+			ev->type = EV_KEY;
+			pressedKey = 0x0;
+		}
+	}
+
+#endif
+
 	if (ev->type != EV_KEY)
 		return;
 		
@@ -581,6 +671,18 @@ void eRCDeviceInputDev::handleCode(long rccode)
 	}
 #endif
 
+#if SX88V2RC
+	// sx88v2 fastforward and rewind are swapped
+	if (ev->code == KEY_FASTFORWARD)
+	{
+		ev->code = KEY_REWIND;
+	}
+	else if (ev->code == KEY_REWIND)
+	{
+		ev->code = KEY_FASTFORWARD;
+	}
+#endif
+
 	}
 	
 	switch (ev->value)
-- 
2.34.1

